kmod/src/data.c

```diff
@@ -1274,10 +1281,31 @@ int scoutfs_data_move_blocks(struct inode *from, u64 from_off,
 
 			to_start = to_iblock + (from_start - from_iblock);
 
-			/* insert the new, fails if it overlaps */
-			ret = scoutfs_ext_insert(sb, &data_ext_ops, &to_args,
-						 to_start, len,
-						 map, ext.flags);
+			if (is_stage) {
+				struct scoutfs_extent to_ext;
+
+				ret = scoutfs_ext_next(sb, &data_ext_ops, &to_args,
+						       to_iblock, 1, &to_ext);
+				if (ret)
+					break;
+
+				if (!scoutfs_ext_inside(to_start, len, &to_ext) ||
+				    !(to_ext.flags & SEF_OFFLINE)) {
+					ret = -EINVAL;
+					break;
+				}
+
+				ret = scoutfs_ext_set(sb, &data_ext_ops, &to_args,
+							 to_start, len,
+							 map, ext.flags);
+				if (ret)
+					break;
+			} else {
+				/* insert the new, fails if it overlaps */
+				ret = scoutfs_ext_insert(sb, &data_ext_ops, &to_args,
+							 to_start, len,
+							 map, ext.flags);
+			}
 			if (ret < 0)
 				break;
 
```

Comment:
(closest line I can comment on, alsjdfkjsafd.)  We still need to turn the err = ext_remove below into using _set to revert back to the offline extent when is_stage, otherwise use the existing _remove undoing for the non-stage case.  They can both still bug_on if they fail :/.

Reply:
ok cool.